inference: remove MAX_TYPE_DEPTH limit, rewrite tmerge

vtjnash opened the pull request:
These arbitrary limits don't seem necessary any more (at least for `make test`), and weren't being used where they would have been needed the most (Tuple) anyways. This also tries to improve the associativity of the algorithm and slowing down the convergence, with an eye towards giving better results when `Nothing` gets joined into a mix of abstract types:

```julia
Core.Compiler.tmerge(Union{Complex{Int32}, Complex{Int64}, Nothing}, Complex{Int16})
`Any` => `Union{Nothing, Complex}`
```

fix #26610
close #23077

vtjnash committed:
inference: remove arbitrary typelimits

`type_too_complex` / `MAX_TYPE_DEPTH` is not well-defined
(this was the last remaining usage)

`_typename(a::Vararg)` does not mean what the author probably thought / intended
(it is the same as ..., and thus was dead code)

This limit is now handled with more surgical precision by `limit_type_size` during recursion,
so we can be more permissive during iteration here.

vtjnash committed:
inference: code movement [NFC]

vtjnash committed:
inference: improve associativity of tmerge

and slow down convergence rate of Tuple and UnionAll types

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

quinnj commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26628#issuecomment-377737365) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5841fb4e859864f77f50621cf32e556e4eebdf45/1e76fc4_vs_025a9ee/report.md). cc @ararslan

vtjnash committed:
optimize: make tmerge fully associative

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26628#issuecomment-378683583) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2a291cc26be765bfa26cc438e94f8b5dfda10720/17f8edf_vs_b64a8e0/report.md). cc @ararslan

vtjnash committed:
Dates: remove anti-pattern of using `Tuple{Union{Nothing, Something}, Int}`

While the new optimizer should handle this better,
the optimal efficient way to handle this is to make the missingness entirely disjoin:
`Union{Nothing, Tuple{Something, Int}}`

And make each `val` variable SSA, for better inference

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26628#issuecomment-378974812) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/64236798e2dcf9895f3a7cc9dd99dfc515da58a5/f2a7dec_vs_8bb878d/report.md). cc @ararslan

vtjnash commented:
great. will merge soon if no objections
